man/im_histgr.3
.TH IM_HISTGR 3 "10 May 1991"
.SH NAME
im_histcum, 
im_histeq, 
im_histgr, 
im_histnD, 
im_histnorm, 
im_histspec, 
im_identity,
im_identity_ushort, 
im_ismonotonic 
\- create, display and process histograms or luts
.SH SYNOPSIS
#include <vips/vips.h>

int im_histgr( in, out, bandno )
.br
IMAGE *in, *out;
.br
int bandno;

int im_histnD( in, out, bins )
.br
IMAGE *in, *out;
.br
int bins;

int im_histcum(hist, lut)
.br
IMAGE *hist, *lut;

int im_histnorm(hist, lut)
.br
IMAGE *hist, *lut;

int im_histeq(hist, lut)
.br
IMAGE *hist, *lut;

int im_histspec(histin, histref, lut)
.br
IMAGE *histin, *histref, *lut;

int im_identity(lut, bands)
.br
IMAGE *lut;
.br
int bands;

int im_identity_ushort(lut, bands, sz)
.br
IMAGE *lut;
.br
int bands;
.br
int sz;

int im_ismonotonic( IMAGE *lut, int *out )

.SH DESCRIPTION
.B im_histgr(3) 
writes the histogram of image in to image out. If bandno is -1,
then out will have the same number of bands as in, and each band of out will
have the histogram of the corresponding band of in. Otherwise, bandno selects
the band of the image for which the histogram will be found, numbering from
zero.

Image in may be either FMTUCHAR or FMTUSHORT. If in is uchar, then out will
have 256 elements, one for each possible pixel value. If in is ushort, then
.B im_histgr(3) 
finds the maximum of in, and outputs a histogram with max + 1
elements. 

For example, suppose you have an image from a 12-bit camera, where
each pixel is in the range [0,4095]. Calling 
.B im_histgr(3) 
for this image will
make a histogram with at most 4096 elements. If the histogram is smaller than
this, then it means that the right hand end of the histogram was all zero, and
has not been generated.

Also check 
.B im_histnD(3) 
below for another way to make histograms.

.B im_histnD(3) 
makes a n-dimensional histogram from an n-band image (1, 2 and 3 bands only).
Because 3D histograms can get very large very quickly, the 
.B bins
parameter sets the length of each dimension, that is, the number of bins the
possible numeric range of the image is divided into.

Unsigned 8 and 16 bit images only.

Use 
.B im_histplot(3) 
to graph the histogram for visualisation. See the separate manpage.

.B im_histcum(3)
forms a cumulative histogram. 

.B im_histnorm(3) 
normalises a histogram. The maximum histogram value becomes equal to the
number of pixels in the histogram. In effect, the histogram 
becomes 'square'. Each channel is normalised separately.

.B im_histeq(3)
takes as input a histogram held by the IMAGE descriptor hist and creates an
unsigned char look up table (held by the IMAGE descriptor lut) which when
applied on the original image, (with histogram held by hist) histogram
equalises it. Histogram equalisation is carried out for each band of hist
individually. 

.B im_histspec(3)
creates a lut for transforming an image with histogram held by histin
according to the pdf (probability density function) of a reference image with
histogram held by histref. histin and histref should have the same number
of bands.  

.B im_identity(3)
creates an look-up-table  with Xsize=256, Ysize=1, Bands=bands, Bbits=BBBYTE,
BandFmt=FMTUCHAR, Type=LUT, which is held by the IMAGE descriptor lut.  The
created image consist a bands linear (ramp) lut and is the basis for building
look-up tables.  

.B im_identity_ushort(3)
creates an look-up-table  with Xsize=sz, Ysize=1, Bands=bands, Bbits=BBSHORT,
BandFmt=FMTUSHORT, Type=LUT, which is held by the IMAGE descriptor lut.  The
created image consist of a linear (ramp) lut and is the basis for building
look-up tables.  

.B im_ismonotonic(3)
sets out to non-zero if the look-up table (or histogram) in lut is monotonic,
that is, if it's slope is always >0.

.SH RETURN VALUE
All functions return 0 on success and -1 on error.
.SH SEE ALSO
im_hist(3), im_hsp(3), im_heq(3), im_maplut(3), im_tone_map(3),
im_histplot(3).
.SH COPYRIGHT
The National Gallery and Birkbeck College, 1991 \- 1994.


NEWS
VIPS changed from 7.10 to 7.12 (not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- IMAGE->Hist is deprecated, use im_history_get() instead
- im_append_Hist() removed
- IMAGE->Bbits is deprecated (now ignored)
- im_region_local() replaced by im_region_buffer()

VIPS enhancements
=================

- new backwards and forwards compatible VIPS file format: it has a new 
  metadata system that efficiently shares and propogates ICC profiles,
  EXIF data, etc. & whatever
- speed ups: new pixel buffer cache saves and reuses computations, uses liboil 
  where possible for a small speedup
- SMP scaling: simplified lock system improves SMP scaling, double-buffered 
  image writes overlap computation 
- lower mem use: new mmap window cache shares file mappings, history buffers 
  share and reuse image history metadata
- built-in support for new image formats: OpenEXR, Analyze7, CSV
- RGB16 and GREY16 image hints
- uses GOption for much better command-line argument handling
- better C++ binding: namespaces, VError derives from std::exception, image
  and number vector constants
- python binding 
- gcc sentinel attributes added
- added GValue as an arg type
- added im_benchmark() and friends for testing
- new functions: 
	im_analyze2vips(), im_benchmark*(), im_buildlut(),
	im_concurrency_get/set(), im_contrast_surface(),
	im_contrast_surface_raw(), im_copy_from(), im_copy_morph(),
	im_cp_desc_array(), im_cp_descv(), im_extract_areabands(),
	im_flood_blob_copy(), im_get_option_group(), im_grid(),
	im_header_exists(), im_header_map(), im_history_get(),
	im_invalidate(), im_isscalar(), im_lineset(), im_linreg(),
	im_meta*(), im_msb(), im_norm_dmask(), im_project(),
	im_raw2vips(), IM_REGION_ADDR_TOPLEFT(), im_tile_cache(),
	im_video_test()


VIPS changed from 7.8 to 7.10 (not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- no longer generates vips-config script, instead it makes .pc files for
  pkg-config ... replace any "vips-config" with "pkg-config vips-7.10"
- origin hint semantics changed ... it now records the position in the output
  image of the origin of the input image
- all float to int casts on pixel values are now floor() rather than rint(),
  for consistency with C casting semantics

VIPS enhancements
=================

- nip reworked for gtk+-2.4, now called nip2
- now based on glib, so VIPS no longer has it's own thread, plugin and data
  structure libraries
- i18n support, although not quite complete yet
- new error message API to make i18n possible (compatibility macros mean the
  old API still works)
- new 'start up VIPS' call, im_init_world(), sets up i18n and threads
  (optional to call it, but i18n won't fully work unless you do)
- C programs now just need "#include <vips/vips.h>", no other includes
  required
- wraps the Pango font rendering library, see im_text()
- new function flag: NOCACHE ... means the result of the call should not be
  cached, useful for video grab functions and paintbox functions
- im_embed() now supports a wider range of embedding styles, including a fast
  edge stretch
- all area operations use the new embed style to expand their input rather 
  than their output, so there are no more black borders on filtered images
- other new functions: im_render(), im_cache(), im_extract_bands(), 
  im_copy_swap(), im_rint(), im_make_xy(), im_init_world(), im_error(),
  im_warn(), im_diag(), im_rank_image(), im_replicate()
- many fixes and improvements to old funcs
- configure and portability improvements

Bug fixes
=========

- all serious bug fixes got back-ported to 7.8, I think


VIPS changes from 7.6 to 7.8 (again, not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- output format options specified in filenames: new prototypes for load and
  save tiff and jpeg
- C++ API name changes ... error => VError
- include path change
    - C includes <vips/vips.h>
    - C++ includes <vips/vipscpp.h>
- im_extract*() numbers bands from zero, not from 1 ... also other funcs which
  refer to bands (eg. im_lrmosaic() etc.)
- mosaic funcs have extra max blend width parameter

VIPS enhancements
=================

- rewritten and much fancier ip (see separate docs), now called nip ... old ip
  (slightly fixed) still there
- mmap windows remove 2GB image size limit
- images have an origin hint field
- icc transforms supported via wrapping of Little CMS
- much, much faster Fourier transforms via wrapping of libfwfft
- ppm/pgm/pbm read/write
- C++ API enhancements: new constructors, more operator equivalences
- new colourspace: Yxy
- support for D50 colour temperature images
- new functions: im_image(), im_system(), im_version*(), im_blend(),
  im_copy_set(), im_flood_blob(), im_icc_*(), im_open_local_array(),
  im_header_*(), im_sign(), im_ceil(), im_floor(), im_remainderconst*(),
  im_print(), im_guess_prefix(), im_remosaic(), im_invertlut(), Yxy funcs,
  color temperature funcs, im_clip2fmt(), im_blend(), im_lab_morph(),
  im_histnorm(), im_histcum(), im_video_v4l(), im_LabS2Lab(), im_Lab2LabS()
- new type: IMAGEVEC
- header is much faster
- ip/nip split to separate packages
- better vips2dj
- better cygwin support, mingw support too
- Mac OS X and Windows build support
- new set of sensible #define names, old names can be turned off with
  -DIM_NO_VIPS7_COMPAT
- many configure improvements

Bug fixes
=========

- speed ups
- fixes to implicit image format conversions 
- im_zoom() spots integer overflow
- im_LabS2LabQ() rounding on a/b slightly broken for a/b == 0
- fixes to refcounting in C++ API
- mask casts in C++ were broken




VIPS Changes for 7.4 to 7.6

Non-backwards compatible changes
================================

- im_histplot() has new (and better) rules
- im_andconst(), im_orconst() and im_eorconst() now take a
  double, not an unsigned char, as their constant argument type

libsrc/histograms_lut/im_histgr.c
/* @(#) im_histgr: make a histogram of an image and saves it into hist.
 * @(#) If input is uchar, output is 256 by 1 image of uint. If input is
 * @(#) ushort, output is max(image) + 1 by 1 image of uint. If bandno is
 * @(#) zero, then output is has same number of bands as input, with each
 * @(#) band being a separate histogram. Otherwise, bandno selects a band
 * @(#) to find the histogram of.
 * @(#) 
 * @(#) Usage:
 * @(#) int im_histgr(image, hist, bandno)
 * @(#) IMAGE *image, *hist;
 * @(#) int bandno;
 * @(#)
 * @(#)  Returns 0 on success and -1 on error
 *
 * Copyright: 1990, 1991, N. Dessipris.
 *
 * Author: Nicos Dessipris.
 * Written on: 09/07/1990
 * Modified on : 11/03/1991
 * 19/7/93 JC
 *	- test for Coding type added
 * 26/10/94 JC
 *	- rewritten for ANSI
 *	- now does USHORT too
 *	- 5 x faster!
 * 2/6/95 JC
 *	- rewritten for partials
 * 3/3/01 JC
 *	- tiny speed ups
 * 21/1/07
 * 	- number bands from zero
 */

/*

    This file is part of VIPS.
    
    VIPS is free software; you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

 */

/*

    These files are distributed with VIPS - http://www.vips.ecs.soton.ac.uk

 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif /*HAVE_CONFIG_H*/
#include <vips/intl.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <vips/vips.h>

#ifdef WITH_DMALLOC
#include <dmalloc.h>
#endif /*WITH_DMALLOC*/

/* Accumulate a histogram in one of these.
 */
typedef struct {
	int bands;		/* Number of bands in output */
	int which;		/* If one band in out, which band of input */
	int size;		/* Length of bins */
	int mx;			/* Maximum value we have seen */
	unsigned int **bins;	/* All the bins! */
} Histogram;

/* Build a Histogram.
 */
static Histogram *
build_hist( IMAGE *out, int bands, int which, int size )
{
	int i;
	Histogram *hist = IM_NEW( out, Histogram );

	if( !hist || !(hist->bins = IM_ARRAY( out, bands, unsigned int * )) )
		return( NULL );
	for( i = 0; i < bands; i++ ) {
		if( !(hist->bins[i] = IM_ARRAY( out, size, unsigned int )) )
			return( NULL );
		memset( hist->bins[i], 0, size * sizeof( unsigned int ) );
	}

	hist->bands = bands;
	hist->which = which;
	hist->size = size;
	hist->mx = 0;

	return( hist );
}

/* Build a sub-hist, based on the main hist.
 */
static void *
build_subhist( IMAGE *out, void *a, void *b )
{
	Histogram *mhist = (Histogram *) a;

	return( (void *) 
		build_hist( out, mhist->bands, mhist->which, mhist->size ) );
}

/* Join a sub-hist onto the main hist.
 */
static int
merge_subhist( void *seq, void *a, void *b )
{
	Histogram *shist = (Histogram *) seq;
	Histogram *mhist = (Histogram *) a;
	int i, j;

	/* Sanity!
	 */
	if( shist->bands != mhist->bands || shist->size != mhist->size )
		error_exit( "sanity failure in merge_subhist" );

	/* Add on sub-data.
	 */
	mhist->mx = IM_MAX( mhist->mx, shist->mx );
	for( i = 0; i < mhist->bands; i++ )
		for( j = 0; j < mhist->size; j++ )
			mhist->bins[i][j] += shist->bins[i][j];

	/* Blank out sub-hist to make sure we can't add it again.
	 */
	shist->mx = 0;
	for( i = 0; i < shist->bands; i++ )
		shist->bins[i] = NULL;
	
	return( 0 );
}

/* Histogram of all bands of a uchar image.
 */
static int
find_uchar_hist( REGION *reg, void *seq, void *a, void *b )
{
	Histogram *hist = (Histogram *) seq;
	Rect *r = &reg->valid;
	IMAGE *im = reg->im;
	int le = r->left;
	int to = r->top;
	int bo = IM_RECT_BOTTOM(r);
	int nb = im->Bands;
	int x, y, z;

	/* Accumulate!
	 */
	for( y = to; y < bo; y++ ) {
		PEL *p = (PEL *) IM_REGION_ADDR( reg, le, y );
		int i;

		for( i = 0, x = 0; x < r->width; x++ )
			for( z = 0; z < nb; z++, i++ )
				hist->bins[z][p[i]]++;
	}

	/* Note the maximum.
	 */
	hist->mx = 255;

	return( 0 );
}

/* Histogram of a selected band of a uchar image.
 */
static int
find_uchar_hist_extract( REGION *reg, void *seq, void *a, void *b )
{
	Histogram *hist = (Histogram *) seq;
	Rect *r = &reg->valid;
	IMAGE *im = reg->im;
	int le = r->left;
	int to = r->top;
	int bo = IM_RECT_BOTTOM(r);
	unsigned int *bins = hist->bins[0];
	int nb = im->Bands;
	int max = r->width * nb;
	int x, y;

	/* Accumulate!
	 */
	for( y = to; y < bo; y++ ) {
		PEL *p = (PEL *) IM_REGION_ADDR( reg, le, y );

		for( x = hist->which; x < max; x += nb ) 
			bins[p[x]]++;
	}

	/* Note the maximum.
	 */
	hist->mx = 255;

	return( 0 );
}

/* Histogram of all bands of a ushort image.
 */
static int
find_ushort_hist( REGION *reg, void *seq, void *a, void *b )
{
	Histogram *hist = (Histogram *) seq;
	Rect *r = &reg->valid;
	IMAGE *im = reg->im;
	int le = r->left;
	int to = r->top;
	int bo = IM_RECT_BOTTOM(r);
	int mx = hist->mx;
	int nb = im->Bands;
	int x, y, z;

	/* Accumulate!
	 */
	for( y = to; y < bo; y++ ) {
		unsigned short *p = (unsigned short *) 
			IM_REGION_ADDR( reg, le, y );
		int i;

		for( i = 0, x = 0; x < r->width; x++ )
			for( z = 0; z < nb; z++, i++ ) {
				int v = p[i];

				/* Adjust maximum.
				 */
				if( v > mx )
					mx = v;

				hist->bins[z][v]++;
			}
	}

	/* Note the maximum.
	 */
	hist->mx = mx;

	return( 0 );
}

/* Histogram of all bands of a ushort image.
 */
static int
find_ushort_hist_extract( REGION *reg, void *seq, void *a, void *b )
{
	Histogram *hist = (Histogram *) seq;
	Rect *r = &reg->valid;
	IMAGE *im = reg->im;
	int le = r->left;
	int to = r->top;
	int bo = IM_RECT_BOTTOM(r);
	int mx = hist->mx;
	unsigned int *bins = hist->bins[0];
	int nb = im->Bands;
	int max = nb * r->width;
	int x, y;

	/* Accumulate!
	 */
	for( y = to; y < bo; y++ ) {
		unsigned short *p = (unsigned short *) 
			IM_REGION_ADDR( reg, le, y ) + hist->which;

		for( x = hist->which; x < max; x += nb ) {
			int v = p[x];

			/* Adjust maximum.
			 */
			if( v > mx )
				mx = v;

			bins[v]++;
		}
	}

	/* Note the maximum.
	 */
	hist->mx = mx;

	return( 0 );
}

int 
im_histgr( IMAGE *in, IMAGE *out, int bandno )
{
	int size;		/* Length of hist */
	int bands;		/* Number of bands in output */
	Histogram *mhist;
	im_generate_fn scanfn;
	int i, j;
	unsigned int *obuffer, *q;

	/* Check images. PIO from in, WIO to out.
	 */
	if( im_pincheck( in ) || im_outcheck( out ) )
		return( -1 );
	if( in->Coding != IM_CODING_NONE ) {
		im_error( "im_histgr", _( "uncoded images only" ) );
		return( -1 );
	}

	/* Find the range of pixel values we must handle.
	 */
	if( in->Bbits == IM_BBITS_BYTE && in->BandFmt == IM_BANDFMT_UCHAR ) 
		size = 256;
	else if( in->Bbits == IM_BBITS_SHORT && 
		in->BandFmt == IM_BANDFMT_USHORT )
		size = 65536;
	else {
		im_error( "im_histgr", _( "input not uchar or ushort" ) );
		return( -1 );
	}

	/* How many output bands?
	 */
	if( bandno > in->Bands || bandno < -1 ) {
		im_error( "im_histgr", _( "bad band parameter" ) );
		return( -1 );
	}
	if( bandno == -1 ) 
		bands = in->Bands;
	else 
		bands = 1;

	/* Build main hist we accumulate data in.
	 */
	if( !(mhist = build_hist( out, bands, bandno, size )) )
		return( -1 );

	/* Select scan function.
	 */
	if( in->BandFmt == IM_BANDFMT_UCHAR && bandno == -1 ) 
		scanfn = find_uchar_hist;
	else if( in->BandFmt == IM_BANDFMT_UCHAR )
		scanfn = find_uchar_hist_extract;
	else if( in->BandFmt == IM_BANDFMT_USHORT && bandno == -1 )
		scanfn = find_ushort_hist;
	else
		scanfn = find_ushort_hist_extract;

	/* Accumulate data.
	 */
	if( im_iterate( in, 
		build_subhist, scanfn, merge_subhist, mhist, NULL ) )
		return( -1 );

	/* Make the output image.
	 */
	if( im_cp_desc( out, in ) ) 
		return( -1 );
	im_initdesc( out,
		mhist->mx + 1, 1, bands, IM_BBITS_INT, IM_BANDFMT_UINT, 
		IM_CODING_NONE, IM_TYPE_HISTOGRAM, 1.0, 1.0, 0, 0 );
	if( im_setupout( out ) )
		return( -1 );

	/* Interleave for output.
	 */
	if( !(obuffer = IM_ARRAY( out, 
		IM_IMAGE_N_ELEMENTS( out ), unsigned int )) )
		return( -1 );
	for( q = obuffer, j = 0; j < out->Xsize; j++ )
		for( i = 0; i < out->Bands; i++ )
			*q++ = mhist->bins[i][j];

	/* Write interleaved buffer into hist.
	 */
	if( im_writeline( 0, out, (PEL *) obuffer ) )
		return( -1 );

	return( 0 );
}
